ydb/core/tx/datashard/sample_k.cpp

```diff
@@ -0,0 +1,331 @@
+        }
+        Response->Record.SetStatus(NKikimrTxDataShard::TEvSampleKResponse::DONE);
+        // TODO(mbkkt) some retry? how?
+        ctx.Send(ResponseActorId, Response.Release());
+    }
+};
+
+class TDataShard::TTxHandleSafeSampleKScan: public NTabletFlatExecutor::TTransactionBase<TDataShard> {
+public:
+    TTxHandleSafeSampleKScan(TDataShard* self, TEvDataShard::TEvSampleKRequest::TPtr&& ev)
+        : TTransactionBase(self)
+        , Ev(std::move(ev))
+    {
+    }
+
+    bool Execute(TTransactionContext&, const TActorContext& ctx) {
+        Self->HandleSafe(Ev, ctx);
+        return true;
+    }
+
+    void Complete(const TActorContext&) {
+        // nothing
+    }
+
+private:
+    TEvDataShard::TEvSampleKRequest::TPtr Ev;
+};
+
+void TDataShard::Handle(TEvDataShard::TEvSampleKRequest::TPtr& ev, const TActorContext&) {
+    Execute(new TTxHandleSafeSampleKScan(this, std::move(ev)));
+}
+
+void TDataShard::HandleSafe(TEvDataShard::TEvSampleKRequest::TPtr& ev, const TActorContext& ctx) {
+    const auto& record = ev->Get()->Record;
+    TRowVersion rowVersion(record.GetSnapshotStep(), record.GetSnapshotTxId());
+
+    // Note: it's very unlikely that we have volatile txs before this snapshot
+    if (VolatileTxManager.HasVolatileTxsAtSnapshot(rowVersion)) {
+        VolatileTxManager.AttachWaitingSnapshotEvent(rowVersion,
+                                                     std::unique_ptr<IEventHandle>(ev.Release()));
+        return;
+    }
+    const ui64 id = record.GetId();
+
+    auto response = MakeHolder<TEvDataShard::TEvSampleKResponse>();
+    response->Record.SetId(id);
+    response->Record.SetTabletId(TabletID());
+
+    TScanRecord::TSeqNo seqNo = {record.GetSeqNoGeneration(), record.GetSeqNoRound()};
+    response->Record.SetRequestSeqNoGeneration(seqNo.Generation);
+    response->Record.SetRequestSeqNoRound(seqNo.Round);
+
+    auto badRequest = [&](const TString& error) {
+        response->Record.SetStatus(NKikimrTxDataShard::TEvSampleKResponse::BAD_REQUEST);
+        auto issue = response->Record.AddIssues();
+        issue->set_severity(NYql::TSeverityIds::S_ERROR);
+        issue->set_message(error);
+        ctx.Send(ev->Sender, std::move(response));
+    };
+
+    if (const ui64 shardId = record.GetTabletId(); shardId != TabletID()) {
+        badRequest(TStringBuilder() << "Wrong shard " << shardId << " this is " << TabletID());
+        return;
+    }
+
+    const TTableId tableId{record.GetOwnerId(), record.GetPathId()};
+    const auto* userTableIt = GetUserTables().FindPtr(tableId.PathId.LocalPathId);
+    if (!userTableIt) {
+        badRequest(TStringBuilder() << "Unknown table id: " << tableId.PathId.LocalPathId);
+        return;
+    }
+    Y_ABORT_UNLESS(*userTableIt);
+    const auto& userTable = **userTableIt;
+
+    if (const auto* recCard = ScanManager.Get(id)) {
+        if (recCard->SeqNo == seqNo) {
+            // do no start one more scan
```

Comment:
This should be covered with a unit test.

Reply:
already covered btw, static sId need because of this check

Reply:
Added